### PR TITLE
Add `Execution` service

### LINKED ISSUE #93: Add `PromptExecutionService`  service
### Description
This service will execute prompt using a broker and save the execution results to `PromptExecutionHistory`

### Requirements

- project must be runnable
- add migrations separately
- add mapping between prompt execution result and execution history
- execution must be parallel

### Deliverables

- `PromptExecutionProcessingService` - processing service responsible for executing the prompt and saving the results
- `PromptExecutionBroker` - broker responsible for actually calling Semantic Kernel with prompt and arguments
- `ExecuteWorkflowSinglePromptEvent` - event to trigger execution service to execute the prompt
- {workflowId:guid}/execute/{promptId:guid}"

### Architecture

Screenshot 2024-03-11 at 13 35 23

### PR DESCRIPTION
**Pull request by isxoqovxasanboy**
closes #93

### CODE REVIEW

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Brokers/IPromptExecutionBroker.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+using Feedback.Analyzer.Domain.Common.Exceptions;
+using Microsoft.SemanticKernel;
+
+namespace Feedback.Analyzer.Application.Common.Prompts.Brokers;
+
+/// <summary>
+/// Interface for prompt execution broker
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Brokers/IPromptExecutionBroker.cs:7]: Documentation : 

"Defines prompt execution broker functionality"

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Services/IPromptExecutionProcessingService.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+using System.Collections.Immutable;
+using Feedback.Analyzer.Domain.Entities;
+
+namespace Feedback.Analyzer.Application.Common.Prompts.Services;
+
+/// <summary>
+/// Represents a service for executing a prompt asynchronously with the provided arguments.
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Services/IPromptExecutionProcessingService.cs:7]: Doc : 

"Defines prompt execution processing service functionality"

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Services/IPromptExecutionProcessingService.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+/// </summary>
+public interface IPromptExecutionProcessingService
+{
+    ValueTask<IImmutableList<PromptExecutionHistory>> ExecuteAsync(
+        Guid promptId,
+        Dictionary<string, string> arguments,
+        uint executionCount = 1,
+        CancellationToken cancellationToken = default
+    );
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Services/IPromptExecutionProcessingService.cs:16]: Add docs

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Workflows/Events/ExecuteWorkflowSinglePromptEvent.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+using Feedback.Analyzer.Domain.Common.Events;
+
+namespace Feedback.Analyzer.Application.Common.Workflows.Events;
+
+public class ExecuteWorkflowSinglePromptEvent : Event
+{
+    public Guid WorkflowId { get; set; }
+    
+    public Guid PromptId { get; set; }
+}
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Workflows/Events/ExecuteWorkflowSinglePromptEvent.cs:10]: Add docs and change to record, change `Event` to record as well, change setters to initializers ( set -> init )

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Brokers/PromptExecutionBroker.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+using Feedback.Analyzer.Application.Common.Prompts.Brokers;
+using Feedback.Analyzer.Domain.Common.Exceptions;
+using Feedback.Analyzer.Domain.Extensions;
+using Microsoft.SemanticKernel;
+
+namespace Feedback.Analyzer.Infrastructure.Common.Prompts.Brokers;
+
+public class PromptExecutionBroker(Kernel kernel) : IPromptExecutionBroker
+{
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Brokers/PromptExecutionBroker.cs:9]: Add type docs  :

"Provides prompt execution broker functionality and mapping arguments and results"

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Services/PromptExecutionProcessingService.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    )
+    {
+        // Query prompt and feedback
+        var prompt = await promptService.GetByIdAsync(promptId, cancellationToken: cancellationToken) ??
+                     throw new InvalidOperationException($"Could not execute prompt, prompt with id {promptId} not found.");
+
+        // Map arguments to kernel arguments
+        var kernelArguments = new KernelArguments();
+        foreach (var keyValuePair in arguments)
+            kernelArguments.Add(keyValuePair.Key, keyValuePair.Value);
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Services/PromptExecutionProcessingService.cs:34]: Move arguments mapping to broker

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Services/PromptExecutionProcessingService.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+using Feedback.Analyzer.Application.PromptsHistory.Services;
+using Feedback.Analyzer.Domain.Common.Commands;
+using Feedback.Analyzer.Domain.Common.Exceptions;
+using Feedback.Analyzer.Domain.Entities;
+using Microsoft.SemanticKernel;
+using Newtonsoft.Json;
+
+namespace Feedback.Analyzer.Infrastructure.Common.Prompts.Services;
+
+public class PromptExecutionProcessingService(
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Services/PromptExecutionProcessingService.cs:14]: Add type docs

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Services/PromptExecutionProcessingService.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            foreach (var history in histories.SkipLast(1))
+                executionHistories.Add(await promptsExecutionHistoryService.CreateAsync(history, createHistoryCommandOptions, cancellationToken));
+        }
+
+        executionHistories.Add(await promptsExecutionHistoryService.CreateAsync(histories.Last(), default, cancellationToken: cancellationToken));
+      
+        return executionHistories.ToImmutableList();
+    }
+
+    private PromptExecutionHistory MapToHistory(Guid promptId, FuncResult<string?> result, double elapsedMilliseconds)
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Prompts/Services/PromptExecutionProcessingService.cs:76]: Add docs 

"Maps prompt execution result to execution history"

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Api/Configurations/HostConfiguration.Extensions.cs [hunk 9]
```diff
@@ -183,9 +186,33 @@ private static WebApplicationBuilder AddFeedbackInfrastructure(this WebApplicati
...
+
+        // Build kernel
+        var kernel = kernelBuilder.Build();
+
+        builder.Services.AddSingleton(kernel);
+
+        return builder;
+    }
 
     private static WebApplicationBuilder AddPromptAnalysisInfrastructure(this WebApplicationBuilder builder)
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Api/Configurations/HostConfiguration.Extensions.cs:211]: Add docs and rename to `AddSemanticAnalysisInfrastructure`

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Workflows/EventHandlers/ExecuteWorkflowSinglePromptEventHandler.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+                                            .ThenInclude(product => product.CustomerFeedbacks)
+                                            .AsSplitQuery()
+                                            .FirstOrDefaultAsync(cancellationToken: cancellationToken) ??
+                       throw new InvalidOperationException(
+                           $"Could not execute prompt, workflow with id {notification.WorkflowId} not found.");
+
+        var arguments = new Dictionary<string, string>
+        {
+            { "productDescription", workflow.Product.Description },
+            { "customerFeedback", workflow.Product.CustomerFeedbacks.First().Comment }
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Infrastructure/Common/Workflows/EventHandlers/ExecuteWorkflowSinglePromptEventHandler.cs:39]: Use constants for dictionary key

File: src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Services/IPromptExecutionProcessingService.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Feedback.Analyzer.Application.Common.Prompts.Services;
+
+/// <summary>
+/// Represents a service for executing a prompt asynchronously with the provided arguments.
+/// </summary>
+public interface IPromptExecutionProcessingService
+{
+    ValueTask<IImmutableList<PromptExecutionHistory>> ExecuteAsync(
+        Guid promptId,
```
Comment (sulton-max) [inline src/Feedback.Analyzer.Backend/Feedback.Analyzer.Application/Common/Prompts/Services/IPromptExecutionProcessingService.cs:12]: Change to prompt model itself

**Review by botirjonabdumannonovv — APPROVED** [review]

**Review by Ilxom-22 — APPROVED** [review]

**Review by sulton-max — DISMISSED** [review]

**Review by Ilxom-22 — APPROVED** [review]

**Review by Ilxom-22 — APPROVED** [review]

**Review by sulton-max — APPROVED** [review]

**Review by rashidovasadbek — APPROVED** [review]

**Pull request merged**